File: app/src/main/java/at/shockbytes/dante/ui/activity/DownloadActivity.java
package at.shockbytes.dante.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.transition.Fade;
import android.transition.Slide;
import android.view.MenuItem;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import at.shockbytes.dante.R;
import at.shockbytes.dante.core.DanteApp;
import at.shockbytes.dante.dagger.AppComponent;
import at.shockbytes.dante.ui.fragment.DownloadBookFragment;
import at.shockbytes.dante.util.AppParams;
import at.shockbytes.dante.util.books.Book;
import at.shockbytes.dante.util.tracking.Tracker;

public class DownloadActivity extends BaseActivity
        implements DownloadBookFragment.OnBookDownloadedListener {

    private static final String ARG_QUERY = "arg_barcode";

    public static Intent newIntent(Context context, String query) {
        return new Intent(context, DownloadActivity.class).putExtra(ARG_QUERY, query);
    }

    @Inject
    protected Tracker tracker;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_download);
        ((DanteApp) getApplication()).getAppComponent().inject(this);
        setupWindow();

        String query = getIntent().getExtras().getString(ARG_QUERY);
        if (query != null) {
            showDownloadFragment(query);
        } else {
            setResult(RESULT_CANCELED);
            supportFinishAfterTransition();
        }
    }

    @Override
    public void injectToGraph(@NotNull AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == android.R.id.home) {
            supportFinishAfterTransition();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBookDownloaded(Book book) {
        tracker.trackOnBookScanned(book);
        forwardToCaller(book.getId(), true);
    }

    @Override
    public void onCancelDownload() {
        tracker.trackOnFoundBookCanceled();
        forwardToCaller(-1, false);
    }

    @Override
    public void onErrorDownload(String reason) {
        tracker.trackOnDownloadError(reason);
    }

    @Override
    public void onCloseOnError() {
        forwardToCaller(-1, false);
    }

    public void colorSystemBars(int actionBarColor, int actionBarTextColor, int statusBarColor, String bookTitle) {

        if (getSupportActionBar() != null) {
            getSupportActionBar().setBackgroundDrawable(new ColorDrawable(actionBarColor));
            Spannable text = new SpannableString(bookTitle);
            text.setSpan(new ForegroundColorSpan(actionBarTextColor), 0, text.length(),
                    Spannable.SPAN_INCLUSIVE_INCLUSIVE);
            getSupportActionBar().setTitle(text);
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setStatusBarColor(statusBarColor);
        }

    }

    private void forwardToCaller(long bookId, boolean isSuccessful) {

        int resultCode = isSuccessful ? RESULT_OK : RESULT_CANCELED;
        setResult(resultCode, new Intent().putExtra(AppParams.EXTRA_BOOK_ID, bookId));
        supportFinishAfterTransition();
    }

    private void showDownloadFragment(String query) {

        DownloadBookFragment fragment = DownloadBookFragment.newInstance(query);
        fragment.setOnBookDownloadedListener(this);

        getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.activity_download_main_content, fragment)
                .commit();
    }

    private void setupWindow() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setEnterTransition(new Slide());
            getWindow().setExitTransition(new Fade());
        }

        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_cancel);
        }
    }

}

File: app/src/main/java/at/shockbytes/dante/ui/activity/DownloadActivity.kt
package at.shockbytes.dante.ui.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import at.shockbytes.dante.R
import at.shockbytes.dante.dagger.AppComponent
import at.shockbytes.dante.ui.fragment.DownloadBookFragment
import at.shockbytes.dante.util.AppParams
import at.shockbytes.dante.util.books.Book
import at.shockbytes.dante.util.tracking.Tracker
import javax.inject.Inject

class DownloadActivity : BackNavigableActivity(), DownloadBookFragment.OnBookDownloadedListener {

    @Inject
    protected lateinit var tracker: Tracker

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_download)
        setResult(Activity.RESULT_CANCELED, Intent()) // To avoid Kotlin exception
        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_cancel)

        val query = intent.extras?.getString(ARG_QUERY)
        showDownloadFragment(query)
    }

    override fun injectToGraph(appComponent: AppComponent) {
        appComponent.inject(this)
    }

    override fun onBookDownloaded(book: Book) {
        tracker.trackOnBookScanned(book)
        forwardToCaller(book.id, true)
    }

    override fun onCancelDownload() {
        tracker.trackOnFoundBookCanceled()
        forwardToCaller(-1, false)
    }

    override fun onErrorDownload(reason: String) {
        tracker.trackOnDownloadError(reason)
    }

    override fun onCloseOnError() {
        forwardToCaller(-1, false)
    }

    fun colorSystemBars(actionBarColor: Int?, actionBarTextColor: Int?,
                        statusBarColor: Int?, bookTitle: String?) {

        // Default initialize if not set
        val abColor = actionBarColor ?: ContextCompat.getColor(this, R.color.colorPrimary)
        val abtColor = actionBarTextColor ?: ContextCompat.getColor(this, android.R.color.white)
        val sbColor = statusBarColor ?: ContextCompat.getColor(this, R.color.colorPrimaryDark)

        supportActionBar?.setBackgroundDrawable(ColorDrawable(abColor))
        val text = SpannableString(bookTitle)
        text.setSpan(ForegroundColorSpan(abtColor), 0, text.length,
                Spannable.SPAN_INCLUSIVE_INCLUSIVE)
        supportActionBar?.title = text

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.statusBarColor = sbColor
        }
    }

    private fun forwardToCaller(bookId: Long, isSuccessful: Boolean) {
        val resultCode = if (isSuccessful) Activity.RESULT_OK else Activity.RESULT_CANCELED
        setResult(resultCode, Intent().putExtra(AppParams.EXTRA_BOOK_ID, bookId))
        supportFinishAfterTransition()
    }

    private fun showDownloadFragment(query: String?) {

        val fragment = DownloadBookFragment.newInstance(query)
                .setOnBookDownloadedListener(this)
        supportFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.activity_download_main_content, fragment)
                .commit()
    }

    companion object {

        private val ARG_QUERY = "arg_barcode"

        fun newIntent(context: Context, query: String): Intent {
            return Intent(context, DownloadActivity::class.java).putExtra(ARG_QUERY, query)
        }
    }

}
